base/io.jl

```diff
@@ -2,12 +2,269 @@
+in which case it should return `0`.
+Where possible, implementations should make sure `readinto!` performs at most one
+blocking reading IO operation, even when it means only filling part of `v`.
+"""
+function readinto! end
+
+"""
+    readall!(io::IO, v::AbstractVector{UInt8})::Int
+
+Read as many bytes as possible from `io` into the beginning of `v`, returning
+the number of bytes read. This function will continue to read until `io` is EOF,
+or `v` has been filled.
+"""
+function readall! end
+
+# TODO: Still not sure about this write buffer trio of functions below.
+# Should they even exist? Their use is probably limited, and they're not
+# that useful to implement derived functions.
+
+"""
+    getwritebuffer(io::IO)::AbstractVector{UInt8}
+
+Get a mutable write buffer users can fill to write to `io`.
+
+The returned vector `v` must have indices `1:length(v)`.
+Users are free to mutate all bytes in the returned buffer.
+Use `consumewrite` `io` to write bytes from the buffer into the `io`.
+"""
+function getwritebuffer end
+
+"""
+    fillwritebuffer(io::IO)::Int
+
+Expand the writable buffer of `io`, returning the number of bytes the buffer was
+expanded.
+If the buffer cannot be expanded, return `0`.
+"""
+function fillwritebuffer end
+
+"""
+    consumewrite(io::IO, n::UInt)::Nothing
+
+Write the first `n` bytes of the write buffer to `io`, and remove these bytes from the write buffer.
+If `n` is larger than the write buffer, throw a `ConsumedTooMuch` error.
+"""
+function consumewrite end
+
+################### Derived methods ############################
+
+eof(io::IO) = isnothing(get_nonempty_reading_buffer(io))
+
+bytesavailable(io::IO) = length(getbuffer(io))
+
+function readavailable(io::IO)
+    ba = bytesavailable(io)::Int
+
+    # Annoyingly, `readavailable` is documented to do I/O if no data has been buffered.
+    if iszero(ba)
+        iszero(something(fillbuffer(io))::Int) && return UInt8[]
+        ba = bytesavailable(io)::Int
+    end
+
+    v = Vector{UInt8}(undef, ba)
+    n = readinto!(io, v)::Int
+    # This should never occur and indicates an implementation error
+    # or race condition or something like that
+    n == ba || throw(NonSpecificIOError("IO were not able to read all its available bytes"))
+    v
+end
+
+function readall!(io::IO, v::AbstractVector{UInt8})
+    n_read = 0
+    vw = view(v, firstindex(v):lastindex(v))
+    while true
+        isempty(vw) && break
+        n = readinto!(io, vw)::Int
+        iszero(n) && break
+        n_read += n
+        vw = view(v, firstindex(v) + n:lastindex(v))
+    end
+    return n_read
+end
+
+# Default implementation assumes buffering, because it's part of the core interface
+# to implement this for non-buffered IOs
+function readinto!(io::IO, v::AbstractVector{UInt8})
+    isempty(v) && return 0
+    buffer = @something get_nonempty_reading_buffer(io) return 0
+    mn = min(length(v), length(buffer))
+    copyto!(v, firstindex(v), buffer, 1, mn)
+    mn
```

Comment:
Ideally, this would somehow fall back to `readbytes!` to work well with existing IO types. For example, you could check `isnothing(buffer) && !eof(io)` (which should only happen for legacy IO types) and then fallback to `readbytes!` in that case.

Reply:
See comment above.